### ISSUE TITLE
Create a higher-level API for common use cases

### ISSUE THREAD

**Issue by nchammas** [ronf/asyncssh#10]
Asynchronous SSH is a must for those of us who want to efficiently interact with hundreds or even thousands of servers at once.

`asyncssh` stands out as the only Python SSH library (as far as I know) that's built from the ground up on the new `asyncio` module in Python 3.4. That's really great.

However, the API currently offered by `asyncssh` is relatively low-level.

For example, SSH concepts like channels and sessions are exposed to the user, even in cases where they don't need to be. And simple tasks like connecting to a remote host and running a command appear to require quite a bit of boilerplate code.

In contrast, consider some common SSH operations done with [Paramiko](http://www.paramiko.org/), a popular, (synchronous) SSH library for Python:

```python
import paramiko

with paramiko.client.SSHClient() as client:
    client.load_system_host_keys()
    client.connect(hostname='localhost')

    stdin, stdout, stderr = client.exec_command(command='cd Desktop; ls')
    print(stdout.read().decode("utf8"), end='')
    exit_status = stdout.channel.recv_exit_status()
    
    with client.open_sftp() as sftp_client:
        desktop_listing = sftp_client.listdir(path="Desktop")
        sftp_client.get(remotepath="remote/file.txt", localpath="Desktop/file.txt")
```

Now, Paramiko's API is not the end-all and be-all of APIs.

For example, the concept of channels also leaks through here for something very common like checking the exit code of your command. And there are several other improvements that can be made to it.

But it serves as an example of what a higher-level API might look like. With the above snippet of code we ran a command, checked various types of output, and even transferred a file.

An API like this is easier to get started with and broadens the appeal of the library.

It would be a great addition if `asyncssh` had a similar, high-level API that users could use for most tasks, while still allowing the more advanced folks to dig deeper when they need to. Of course, there will be special considerations to make when designing such an API due to the asynchronous nature of the library, but I think it can be done.

Does offering such a higher-level API fit into `asyncssh`'s roadmap?

**Comment by nchammas** [comment]
Here's another example of a higher-level API that `asyncssh` could offer.

Consider the following snippet of code from the [parallel-ssh project README](https://github.com/pkittenis/parallel-ssh#usage-example). `parallel-ssh`, like `asyncssh`, offer asynchronous SSH, but through the use of `gevent` (which currently limits the project to Python 2).

```python
>>> from pssh import ParallelSSHClient
>>> hosts = ['myhost1', 'myhost2']
>>> client = ParallelSSHClient(hosts)
>>> output = client.run_command('ls -ltrh /tmp/', sudo=True)
>>> print output
{'myhost1': {'exit_code': 0, 'stdout': <generator>, 'stderr': <generator>, 'channel': <channel>, 'cmd' : <greenlet>},
 'myhost2': {'exit_code': 0, 'stdout': <generator>, 'stderr': <generator>, 'channel': <channel>, 'cmd' : <greenlet>}}
```

I could see `asyncssh` offering a similar API to allow users to run commands in parallel against a large number of servers at once.

The beautiful thing here is that the user does not even need to know about `asyncio`, coroutines, or any of that stuff to use this API. It can all be encapsulated in `ParallelSSHClient`, or something similar.

This is really empowering for users who are not comfortable with asynchronous programming but still want (and IMO should be able) to reap the benefits.

**Comment by yosemitesammy217** [comment]
Write it and submit merge requests

**Comment by nchammas** [comment]
Sure, I would love to help make this a reality.

But I wouldn't want to charge ahead with contributing something fairly large like this without knowing that, at some level, the project maintainers a) thought it fit in with the project and b) were willing to give guidance on refining any implementation.

Anyway, do you have any feedback on the proposal @yosemitesammy217? (I'm not sure whether "write it and submit a PR" is an endorsement or a challenge. :confused:)

**Comment by yosemitesammy217** [comment]
Treat it as an endorsement.. :-)   Sorry for being a bit terse.  I too would like to see some of the more refined features of paramiko come to asyncssh.  Invoke_shell() specifically.

**Comment by ronf** [comment]
The main boiler plate that AsyncSSH currently adds is the requirement that the caller at some point run the asyncio event loop explicitly, whereas programs like Paramiko hide this from the caller by creating separate threads to do the I/O, which defeats much of the point of using asyncio as you now have to worry about concurrency and can't assume run-to-completion semantics on the callbacks you receive. While I'm very much in favor of the idea of keeping basic tasks in AsyncSSH as simple as possible from an API perspective, I'm not quite sure how to reconcile this particular point.

That said, I'd be happy to explore this idea further, if anyone has any thoughts on how one can stay true to asyncio's execution model.

Regarding invoke_shell, you should be able to do that by leaving out the command argument in the call to create_session(). If you don't pass a command or subsystem, it invokes a shell on the remote machine, and you can send a series of commands by doing writes to the channel. Is there something I'm missing about what paramiko does in this case?

**Comment by nchammas** [comment]
OK, no worries @yosemitesammy217. :)

Thanks for the feedback @ronf. I admit I'm still new to asynchronous programming in general, and have been getting up to speed these past few weeks on the various libraries and frameworks out there.

So you're probably right that naively trying to mimic Paramiko's API, for example, wouldn't work due to the fundamentally different execution models. At some level the user *has* to be aware that they are running coroutines, at least in the case when they want to manage individual, asynchronous SSH calls.

What do you think, then, of offering a single abstraction for interacting with a set of hosts at once, like in the `ParallelSSHClient` example I gave above?

This actually is probably more valuable to the average user than making calls to individual hosts; if you're interested in asynchronous SSH, it's almost certainly because you want to interact with many hosts at once. (Or am I wrong about that? It's true in my case, at least!)

With something like `ParallelSSHClient`, users don't have to worry or even know about the event loop or coroutines. It's all managed within the client. (I think this means that the user cannot have an event loop running elsewhere when they use this client, but I think that's a fair tradeoff.)

All a user would need to know to use `ParallelSSHClient` is:
1. the set of hosts they want to interact with
2. any SSH configs they want to set
3. the inputs they want to provide to the hosts (e.g. commands, files), and
4. the outputs they want to get back (e.g. stdout, stderr, return codes)

What do you think of offering such an abstraction?

**Comment by ronf** [comment]
I could definitely see it making sense to build an application on top of AsyncSSH which provided functionality similar to ParallelSSHClient, and for that application I could see it making sense to say that when you instantiate the client object it creates is own private event loop which it manages for you. Running a common set of commands on multiple hosts at once seems like a somewhat specialized requirement, though, so I'm not sure if it makes sense to build the functionality into AsyncSSH itself.

In terms of the motivation for building AsyncSSH in the first place, my need was actually for a shared SSH server with lots of clients simultaneously connecting to it (basically a multi-user chat system). I also wanted this same server to support other protocols like Telnet and SSL in a common event loop. However, as I built it, I realized I could provide both a client & server without a whole lot of extra effort.

With the current AsyncSSH API, I think it would be pretty straightforward to create multiple SSHClientConnection objects that all shared a common event loop, and even shared a common set of client keys, passwords, or other SSH configuration if that made sense. What you did with each of the sets of output which came back would be application-specific, though, so I'm not sure there's much I could do in AsyncSSH itself to make that part simpler.

Regarding SFTP, that is definitely one thing I haven't tackled yet. I know Paramiko and some of the other toolkits have support for this, and it looks like it's a fairly big piece of additional work. I would not be opposed to having that functionality included in AsyncSSH, but I'm not sure it's something I'll get to myself right away.

**Comment by nchammas** [comment]
OK, this all seems reasonable. Given that it doesn't seem like a higher-level API than the ones already offered belongs in this project, I'll close this issue.

Thanks for taking the time to explain things.

**Comment by nchammas** [comment]
@ronf Does it make sense to open a separate issue for SFTP support, so people can watch it for updates when/if it is implemented?

**Comment by ronf** [comment]
Sure, that sounds like a good idea.

**Comment by nchammas** [comment]
OK, done in #11.
